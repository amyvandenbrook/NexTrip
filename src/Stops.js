import React from 'react'

export const ShowStops = (props) => {
    if (props.stopLoaded) {
        return (
            <ul>
            {props.stopData.map(stop => (
                <li key={stop.Value}>{stop.Text}</li>
            ))}
            </ul>
        )
    } else {
        return (
            <div>Please choose a route above to see related stops.</div>
        )
    }
}