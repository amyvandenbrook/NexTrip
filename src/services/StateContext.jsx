import React, { createContext, useReducer } from 'react'
import Reducer from './Reducer.jsx'

const StateContext = createContext()

const StateProvider = ({ children, initialState }) => {
    const defaultState = {
        routeData: [],
        route: undefined,
        routeChanged: false,
        directionData: [],
        direction: '',
        stopData: [],
    }
    const [state, dispatch] = useReducer(Reducer, {
        ...defaultState,
        ...initialState,
    })

    return (
        <StateContext.Provider
            value={{
                ...state,
                dispatch,
            }}
        >
            {children}
        </StateContext.Provider>
    )
}

export { StateContext, StateProvider }
