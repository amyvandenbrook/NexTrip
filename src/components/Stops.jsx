import React, { useContext, useEffect } from 'react'
import { StateContext } from '../services/StateContext.jsx'
import { getStops } from '../services/AxiosService.jsx'

const Stops = (props) => {
    const { dispatch, route, direction, stopData } = useContext(StateContext)

    useEffect(() => {
        if (route && direction) {
            getStops(route, direction).then((result) => {
                console.log(result)
                dispatch({ type: 'set_stop_data', payload: result.data })
            })
        }
    }, [dispatch, route, direction])

    return stopData.length > 0 ? (
        <>
            <h2>Stops</h2>
            <ul>
                {stopData.map((stop, index) => (
                    <li key={index}>{stop.Text}</li>
                ))}
            </ul>
        </>
    ) : null
}

export default Stops
