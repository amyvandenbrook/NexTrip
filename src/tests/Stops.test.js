import React from 'react'
import { act, create } from 'react-test-renderer'
import { render, fireEvent, screen } from '@testing-library/react'
import axios from 'axios'
import { StateProvider } from '../services/StateContext.jsx'
import Stops from '../components/Stops.jsx'

jest.mock('axios')
axios.get = jest.fn()

describe('Stops', () => {
    let stops
    let component

    test('returns null if no route or direction exists', () => {
        act(() => {
            component = create(
                <StateProvider>
                    <Stops />
                </StateProvider>
            )
        })
        stops = component.toJSON()
        expect(stops).toStrictEqual(null)
    })

    test('requests and displays stops if directions & route exists', async () => {
        const mockReturnedData = [
            { Text: 'mockStop1' },
            { Text: 'mockStop2' },
            { Text: 'mockStop3' },
        ]
        axios.get.mockImplementationOnce(() =>
            Promise.resolve({
                status: 200,
                data: mockReturnedData,
            })
        )
        await act(() => {
            component = create(
                <StateProvider
                    initialState={{
                        route: 'mockRoute',
                        direction: 'mockDirection',
                    }}
                >
                    <Stops />
                </StateProvider>
            )
        })
        stops = component.toJSON()
        expect(stops).toMatchSnapshot()
    })
})
