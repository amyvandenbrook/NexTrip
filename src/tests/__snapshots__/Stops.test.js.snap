// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stops requests and displays stops if directions & route exists 1`] = `
<div
  className="stops"
>
  <h2
    className="header--2"
  >
    Stops
  </h2>
  <ul>
    <li>
      mockStop1
    </li>
    <li>
      mockStop2
    </li>
    <li>
      mockStop3
    </li>
  </ul>
</div>
`;
